viewing_party/party.py

```diff
@@ -1,23 +1,137 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    new_movie = {}
+
+    if title and genre and rating:
+        new_movie["title"] = title
+        new_movie["genre"] = genre
+        new_movie["rating"] = rating
+        return new_movie
+    return None
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if title == movie["title"]:
+            add_to_watched(user_data, movie)
+            user_data["watchlist"].remove(movie)
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    avg_rating = 0.0
+    summed_ratings = 0
+    collected_ratings = []
+
+    if len(user_data["watched"]) >= 1:
+        collected_ratings = [movie["rating"] for movie in user_data["watched"]]
+    else:
+        return avg_rating
+
+    for rating in collected_ratings:
+        summed_ratings += rating
+
+    avg_rating = (summed_ratings / len(collected_ratings))
+    return avg_rating
+
+def get_most_watched_genre(user_data):
+    genre_count = {}
+    top_genre = (0, 0)
 
+    if len(user_data["watched"]) < 1:
+        return None
+
+    for movie in user_data["watched"]:
+        if not movie["genre"] in genre_count:
+            genre_count[movie["genre"]] = 1
+        else:
+            genre_count[movie["genre"]] += 1
+
+    for genre_pair in genre_count.items():
+        if genre_pair[1] > top_genre[1]:
+            top_genre = genre_pair
+
+    return top_genre[0]
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    friends_watched = [movie for friend in user_data["friends"] 
+        for movie in friend["watched"]]
+
+    unique_watched = [movie for movie in user_data["watched"] 
+        if not movie in friends_watched]
+
+    return unique_watched
+
+def get_friends_unique_watched(user_data):
+    unique_watched = []
+
+    friends_watched = [movie for friend in user_data["friends"] 
+        for movie in friend["watched"]]
 
+    friends_watched_dict = {movie["title"]:movie for movie in friends_watched}
+
+    user_titles_watched = [movie["title"] for movie in user_data["watched"]]
+
+    unique_watched = [friends_watched_dict[movie_title] for movie_title in friends_watched_dict.keys() 
+        if not movie_title in user_titles_watched]

```

Comment:
OK, this works as a way to remove duplicates from `friends_watched`, but definitely feels like a bit of complication to really hold on to using list and dictionary comprehensions. Which great using them, I just think it's a bit more on the showing off side here, and impacts how easy this function is to understand.

For example, another way to remove duplicates could have been done in an explicit nested `for` loop:

```python
friends_watched = []
for friend in user_data["friends"]:
    for movie in friend["watched"]:
        if movie not in friends_watched:
            friends_watched.append(movie)

return [movie for movie in user_data["watched"] if movie not in friends_watched]
```

You might notice that this is worse time complexity than your implementation, but it makes up with a slight benefit in space complexity.   I can share the details of why I think that in 1:1 if you want, but essentially you're going to need space for all of the friends' movies, while I only need space for the unique ones.

More valuable to me, I think my version is more readable and thus would be easier to maintain. But there's a lot of different tradeoffs here, and no real certain opinion. (And this is rather minor...)